casper/network.py

```diff
@@ -10,74 +11,62 @@ def __init__(self, validator_set, protocol=BlockchainProtocol, force_justify_mes
             self._collect_initial_messages(),
             protocol.initial_message(None)
         )
+        self.message_queues = {
+            validator: PriorityQueue()
+            for validator in self.validator_set
+        }
+        self.time = 0
 
         self.force_justify_messages = force_justify_messages
 
-    def propagate_message_to_validator(self, message, validator):
-        """Propagate a message to a validator."""
-        assert message.hash in self.global_view.justified_messages, (
-            "...expected only to propagate messages "
-            "from the global view")
-        assert validator in self.validator_set, "...expected a known validator"
-
-        validator.receive_messages(set([message]))
-
-        # HACK TO SEND ALL MESSAGES BEFORE NETWORK REWORK
-        if self.force_justify_messages:
-            self._propagate_messages_needed_to_justify(message, validator)
-
-    def get_message_from_validator(self, validator):
-        """Get a message from a validator."""
-        assert validator in self.validator_set, "...expected a known validator"
-
-        new_message = validator.make_new_message()
-        self.global_view.add_messages(set([new_message]))
-        assert new_message.hash in self.global_view.justified_messages
-
-        return new_message
-
-    def view_initialization(self, view):
-        """Initalizes all validators with all messages in some view."""
-        messages = view.justified_messages.values()
-
-        self.global_view.add_messages(messages)
-
+    #
+    # async network rework
+    #
+    def delay(self, sender, receiver):
+        '''Must be defined in child class.
+        Returns delay of next message for sender to receiver'''
+        raise NotImplementedError
+
+    def send(self, validator, message):
+        self.global_view.add_messages(
+            set([message])
+        )
+        self.message_queues[validator].put((
+            self.time + self.delay(message.sender, validator),
+            message
+        ))
+
+    def receive(self, validator):
+        queue = self.message_queues[validator]
+        if queue.qsize() == 0:
+            return None
+        if queue.peek()[0] > self.time:
+            return None
+
+        return queue.get()[1]
+
+    def receive_all(self, validator):
+        messages = []
+        message = self.receive(validator)
+        while message:
+            messages.append(message)
+            message = self.receive(validator)
+
+        return messages
+
+    def send_to_all(self, message):
```

Comment:
I think it also would be cool to allow for a `send_to_subset` and `receive_subset`.  Could be relatively easy to implement a flat or gaussian random sample from the validator set. 

This would probably be un-needed complexity-- just a thought.

Reply:
`send_to_all` is just a helper because simulations for now need to propagate everything to everyone (even if someone doesn't get the message until way later). A simulation that modeled a gossip network where each validator had a set of peers could use `send_to_subset` to send to their peers, but could also just use `send`.  I'd rather hold off on adding that function to the API until we need/want it for a particular type of simulation